File: Programmars/Level_2/방문 길이.py
def solution(dirs):
    answer = 0
    dy = [1, 0, -1, 0]
    dx = [0, 1, 0, -1]
    command = ['U', 'R', 'D', 'L']
    now = (0, 0, 0, 0)
    path = []
    for dir in dirs:
        next_path = (now[2], now[3], now[2] +
                     dy[command.index(dir)], now[3]+dx[command.index(dir)])
        if -5 <= next_path[2] <= 5 and -5 <= next_path[3] <= 5:
            if not next_path in path:
                answer += 1
                path.append(next_path)
                path.append((next_path[2], next_path[3],
                             next_path[0], next_path[1]))
            now = next_path
    return answer


print(solution("ULURRDLLU"))  # 7
print(solution("LULLLLLLU"))  # 7

File: JavaScript/Programmars/썸머윈터 코딩/방문 길이.js
// 문제: https://programmers.co.kr/learn/courses/30/lessons/49994
// 일부 오답
const solution = (dirs) => {
  const move = {
    U: [-1, 0],
    D: [1, 0],
    R: [0, 1],
    L: [0, -1],
  };
  const isRightPosition = (x, y) => {
    return x >= 0 && x <= 10 && y >= 0 && y <= 10;
  };

  const check = Array.from(Array(11), () => Array(11).fill(0));
  let answer = 0;
  let currentX = 5;
  let currentY = 5;
  dirs.split('').forEach((dir) => {
    const [x, y] = move[dir];
    if (isRightPosition(x + currentX, y + currentY)) {
      currentX += x;
      currentY += y;
      if (check[currentX][currentY] === 0) {
        answer++;
        check[currentX][currentY] = 1;
      }
    }
  });
  console.table(check);

  return answer;
};

console.log(solution('ULURRDLLU')); // 7
// console.log(solution('LULLLLLLU')); // 7
